vsintegration/src/FSharp.Editor/CodeLens/CodeLens.fs

```diff
@@ -0,0 +1,378 @@
+
+open System
+open System.Windows.Controls
+open System.Windows.Media
+open Microsoft.VisualStudio.Text
+open Microsoft.VisualStudio.Text.Editor
+open Microsoft.VisualStudio.Text.Formatting
+open System.ComponentModel.Composition
+open Microsoft.VisualStudio.Utilities
+open Microsoft.CodeAnalysis
+open System.Threading
+open Microsoft.FSharp.Compiler.SourceCodeServices
+open Microsoft.VisualStudio.Shell
+open Microsoft.VisualStudio
+open Microsoft.VisualStudio.LanguageServices
+open System.Windows
+open System.Collections.Generic
+open Microsoft.FSharp.Compiler.Range
+open Microsoft.FSharp.Compiler
+open Microsoft.CodeAnalysis.Editor.Shared.Extensions
+open Microsoft.CodeAnalysis.Editor.Shared.Utilities
+open Microsoft.CodeAnalysis.Classification
+open Internal.Utilities.StructuredFormat
+open Microsoft.VisualStudio.Text.Tagging
+open System.Collections.Concurrent
+open System.Collections
+open System.Windows.Media.Animation
+
+open Microsoft.VisualStudio.FSharp.Editor.Logging
+
+type CodeLensTag(width, topSpace, baseline, textHeight, bottomSpace, affinity, tag:obj, providerTag:obj) =
+    inherit SpaceNegotiatingAdornmentTag(width, topSpace, baseline, textHeight, bottomSpace, affinity, tag, providerTag)
+
+type internal CodeLens =
+    { TaggedText: Async<(ResizeArray<Layout.TaggedText> * QuickInfoNavigation) option> 
+      mutable Computed: bool }
+
+type internal CodeLensTagger  
+    (
+        workspace: Workspace, 
+        documentId: Lazy<DocumentId>,
+        buffer: ITextBuffer, 
+        checker: FSharpChecker,
+        projectInfoManager: ProjectInfoManager,
+        typeMap: Lazy<ClassificationTypeMap>,
+        gotoDefinitionService: FSharpGoToDefinitionService
+     ) as self =
+    
+    let tagsChanged = new Event<EventHandler<SnapshotSpanEventArgs>,SnapshotSpanEventArgs>()
+    let formatMap = lazy typeMap.Value.ClassificationFormatMapService.GetClassificationFormatMap "tooltip"
+    let codeLensResults = ConcurrentDictionary<string, CodeLens>()
+    let codeLensUIElementCache = ConcurrentDictionary()
+    let currentVisibleLineNumbers = ConcurrentDictionary()
+    let mutable firstTimeChecked = false
+    let mutable bufferChangedCts = new CancellationTokenSource()
+    let mutable layoutChangedCts = new CancellationTokenSource()
+    let mutable view: IWpfTextView option = None
+    let mutable codeLensLayer: IAdornmentLayer option = None
+    let mutable recentFirstVsblLineNmbr, recentLastVsblLineNmbr = 0, 0
+    
+    let layoutTagToFormatting (layoutTag: LayoutTag) =
+        layoutTag
+        |> RoslynHelpers.roslynTag
+        |> ClassificationTags.GetClassificationTypeName
+        |> typeMap.Value.GetClassificationType
+        |> formatMap.Value.GetTextProperties   
+
+    let executeCodeLenseAsync () =  
+        let uiContext = SynchronizationContext.Current
+        asyncMaybe {
+            let! view = view
+            do! Async.Sleep 800 |> liftAsync
+            logInfof "Rechecking code due to buffer edit!"
```

Comment:
add a `[CodeLens]` prefix to these messages will make it clear which feature they're coming from when logging is enabled